_pytest/assertion/rewrite.py

```diff
@@ -144,28 +145,25 @@ def _should_rewrite(self, name, fn_pypath, state):
         if fn_pypath.basename == 'conftest.py':
             state.trace("rewriting conftest file: %r" % (fn,))
             return True
-        elif self.session is not None:
+
+        if self.session is not None:
             if self.session.isinitpath(fn):
                 state.trace("matched test file (was specified on cmdline): %r" %
                             (fn,))
                 return True
-            else:
-                # modules not passed explicitly on the command line are only
-                # rewritten if they match the naming convention for test files
-                session = self.session  # avoid a cycle here
-                self.session = None
-                try:
-                    for pat in self.fnpats:
-                        if fn_pypath.fnmatch(pat):
-                            state.trace("matched test file %r" % (fn,))
-                            return True
-                finally:
-                    self.session = session
-                    del session
-        else:
-            for marked in self._must_rewrite:
-                if marked.startswith(name):
-                    return True
+
+        # modules not passed explicitly on the command line are only
+        # rewritten if they match the naming convention for test files
+        for pat in self.fnpats:
+            if fnmatch(fn_pypath.basename, pat):
+                state.trace("matched test file %r" % (fn,))
+                return True
+
+        for marked in self._must_rewrite:
+            if marked.startswith(name):
```

Comment:
However thinking about it some more maybe `name.startswith(marked)` would be better behaviour.  We should probably also update the docstring in `assertion/__init__.py;register_assert_rewrite` to read "will make sure that this module or all modules inside the package will get their assert statements rewritten" in that case though.

The only downside of that way round that I can think of is that someone may want to register `mypkg.__init__` explicitly rather then just `mypkg`, but that's probably ok.